### PR TITLE
Product creation AI: Show confirm alert when trying to dismiss by dragging

### LINKED ISSUE #10799: Product Creation AI: Show confirm alert when trying to dismiss by dragging

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #10799 

## Description

This PR intercepts the swipe-to-dismiss gesture on the product creation AI sheet. A confirm action sheet is displayed if there are any changes to product name or description.

## Testing instructions

- Log in to a WPCom or self-hosted store with Jetpack AI installed.
- Switch to the Products tab and select "+" to add a new product.
- Select add product with AI.
- Follow the steps to add product name, features and generate details.
- Try swiping-to-dismiss the sheet, an action sheet should be displayed asking to confirm the dismissal.
- Navigate back to previous screen, the same sheet should be displayed when attempting to swipe to dismiss.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/5533851/fe66f6c3-08f3-43f6-baba-130f02473200

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Container/AddProductWithAIContainerViewModel.swift [hunk 1]
```diff
@@ -23,6 +23,10 @@ final class AddProductWithAIContainerViewModel: ObservableObject {
     let siteID: Int64
     let source: AddProductCoordinator.Source
 
+    var canBeDismissed: Bool {
+        productName.isEmpty && productFeatures.isEmpty && productDescription == nil
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Container/AddProductWithAIContainerViewModel.swift]: Nit: Consider adding unit tests for this.

**Review by selanthiraiyan — COMMENTED** [review]
I think there exists an edge case in which navigating back after deleting the features and name text field still shows the confirmation alert. 

i.e. The name and features fields are empty, but the confirmation alter is displayed. 

https://github.com/woocommerce/woocommerce-ios/assets/524475/13074d5c-97b7-4338-8cea-cb33729af133

**Comment by itsmeichigo** [comment]
Thanks for the review @selanthiraiyan! To check the state of the product name screen, I've moved its view model to the container view's view model. I updated the check and logic of both screens in the subsequent commits - this PR is ready for another look.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift [hunk 2]
```diff
@@ -2,6 +2,45 @@ import XCTest
...
+
+    func test_canBeDismissed_returns_true_if_current_step_is_productName_and_the_name_field_is_empty() {
+        // Given
+        let viewModel = AddProductWithAIContainerViewModel(siteID: 123, source: .productsTab, onCancel: {}, onCompletion: { _ in })
+
+        // Then
+        XCTAssertTrue(viewModel.canBeDismissed)
+    }
+
+    func test_canBeDismissed_returns_False_if_current_step_is_productName_and_the_name_field_is_not_empty() {
```
Comment (selanthiraiyan) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift]: Super nit:
```suggestion
    func test_canBeDismissed_returns_false_if_current_step_is_productName_and_the_name_field_is_not_empty() {
```

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift [hunk 3]
```diff
@@ -2,6 +2,45 @@ import XCTest
...
+        let viewModel = AddProductWithAIContainerViewModel(siteID: 123, source: .productsTab, onCancel: {}, onCompletion: { _ in })
+
+        // When
+        viewModel.addProductNameViewModel.productNameContent = "iPhone 15"
+
+        // Then
+        XCTAssertFalse(viewModel.canBeDismissed)
+    }
+
+    func test_canBeDismissed_returns_False_if_current_step_is_not_product_name() {
```
Comment (selanthiraiyan) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift]: Super nit:
```suggestion
    func test_canBeDismissed_returns_false_if_current_step_is_not_product_name() {
```

**Review by selanthiraiyan — APPROVED** [review]
Thank you for the quick updates. LGTM. 🚀
